geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/AvroIngestProcessor.scala

```diff
@@ -143,41 +99,158 @@ trait AvroIngestProcessor extends FeatureTypeProcessor {
+                      s"use ${FeatureNameOverride.getName} instead")
+                old
+              } else {
+                logger.warn(
+                  s"Ignoring property ${Properties.AvroSftName.getName}: ${old.get} " +
+                      s"in favor of ${FeatureNameOverride.getName}: ${fno.get}")
+                fno
+              }
+            }
+
+            val sft = nameArg match {
+              case Some(name) if name != reader.getSft.getTypeName => SimpleFeatureTypes.renameSft(reader.getSft, name)
+              case _ => reader.getSft
             }
-            features.foreach { sf =>
-              try {
-                if (useProvidedFid) {
-                  sf.getUserData.put(Hints.USE_PROVIDED_FID, java.lang.Boolean.TRUE)
+
+            val mapper = checkSchemaAndMapping(sft)
+            val features = mapper.map(m => reader.map(m.apply)).getOrElse(reader)
+            val writer = writers.borrowWriter(sft.getTypeName)
+            try {
+              features.foreach { sf =>
+                try {
+                  if (useProvidedFid) {
+                    sf.getUserData.put(Hints.USE_PROVIDED_FID, java.lang.Boolean.TRUE)
+                  }
+                  writer.apply(sf)
+                  success += 1L
+                } catch {
+                  case NonFatal(e) =>
+                    failure += 1L
+                    logError(sf, e)
                 }
-                writer.apply(sf)
-                success += 1L
-              } catch {
-                case NonFatal(e) =>
-                  failure += 1L
-                  logError(sf, e)
               }
+            } finally {
+              writers.returnWriter(writer)
             }
           } finally {
             reader.close()
           }
         }
       })
 
-      (success, failure)
+      IngestResult(success, failure)
     }
 
-    def buildConverter(input: SimpleFeatureType, output: SimpleFeatureType): Option[SimpleFeature => SimpleFeature] = {
-      FeatureTypeProcessor.checkCompatibleSchema(input, output).map { error =>
-        if (matchMode == LenientMatch) {
-          convert(input, output)
-        } else {
-          throw error
-        }
+    private def checkSchemaAndMapping(sft: SimpleFeatureType): Option[SimpleFeature => SimpleFeature] = {
+      val existing = Try(store.getSchema(sft.getTypeName)).getOrElse(null)
+
+      if (existing == null) {
+        logger.info(s"Creating schema ${sft.getTypeName}: ${SimpleFeatureTypes.encodeType(sft)}")
+        store.createSchema(sft)
+        None
+      } else if (SimpleFeatureTypes.compare(sft, existing) == 0) {
+        None
+      } else if (mode == CompatibilityMode.Update) {
+        logger.info(
+          s"Updating schema ${sft.getTypeName}:" +
+              s"\n  from: ${SimpleFeatureTypes.encodeType(existing)}" +
+              s"\n  to:   ${SimpleFeatureTypes.encodeType(sft)}")
+        store.updateSchema(sft.getTypeName, sft)
```

Comment:
Wow, NiFi can call updateSchema?!  

Neat and kinda crazy/awesome.

Reply:
UpdateSchema gets a Curator distributed lock, right?

Reply:
yeah it's had that option for a while now. most stores will get a lock, not 100% sure all of them will...